### ISSUE TITLE
RuntimeError: Set changed size during iteration

### ISSUE THREAD

**Issue by abjose** [Cronenbrogues/cronenbroguelike#77]
Running into this one sometimes while running a randomized player for a long time. Very possible this is an artifact of the way I'm randomly choosing actions.

```
> look

Smoker puffs mellowly on a cigarette, looking extremely fly.

An idol with ruby eyes and a soot-stained maw sneers at you, showing carven fangs. Directly beneath its
chin sits an altar, resting atop a stone plinth. Its sides are embellished with bas-relief ivory friezes
depicting various acts both lewd and violent. On one side is depicted the apparent sacrifice of a man
bearing a sprig of leaves, overseen by an angry beast. On another is shown a burnt offering of the same
leaves. The smoke rises up toward a face, whose mouth gapes in a slack-jawed smile.

Thick wreaths of acrid smoke hang in the air.

A smoker stands meditatively, cigarette in hand. He looks cool, maybe cooler than you.

Exits are through a murky tunnel.

The smoke whirls for a moment on the altar as though set there for an offering. The idol's ruby eyes
seem to blaze as the smoke makes languorous curls beneath its nose. With a sound like the clatter of
gravel and like the snuffling of many beasts, the idol draws the smoke into its nostrils. Its mouth
shudders and falls open, smashing the altar beneath to shards.

All beings present are pelted with debris.

Health damaged by 1.

You die.

...

You recall your death by pelting with stone fragments. The memory fades away.

You know only that you have been here for interminable years, that you have died innumerable times, and
that someone once told you there was a way out. You were told this an eon ago, or maybe a day, but the
stubborn hope of escape glisters in your mind.

A desk is strewn with sheaves of paper. Little of sense is written there.

Exits are north, south, west.
```
(crashes)
``` python
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Downloads/cronenbroguelike/cronenbroguelike/__main__.py", line 24, in <module>
    game.main()
  File "/Downloads/cronenbroguelike/cronenbroguelike/game.py", line 111, in main
    adventurelib._handle_command("random")
  File "/.local/lib/python3.6/site-packages/adventurelib.py", line 600, in _handle_command
    func(**args)
  File "<string>", line 4, in wrapped
  File "/Downloads/cronenbroguelike/cronenbroguelike/commands.py", line 90, in random_action
    adventurelib._handle_command(final_command)
  File "/.local/lib/python3.6/site-packages/adventurelib.py", line 600, in _handle_command
    func(**args)
  File "<string>", line 4, in wrapped
  File "/usr/lib/python3.6/contextlib.py", line 88, in __exit__
    next(self.gen)
  File "/Downloads/cronenbroguelike/engine/globals.py", line 85, in poll_events
    event.execute()
  File "/Downloads/cronenbroguelike/cronenbroguelike/rooms.py", line 152, in execute
    for character in self.room.characters:
RuntimeError: Set changed size during iteration
```

**Comment by flosincapite** [comment]
No, this is a real bug. It's another artifact of the way resetting the game currently works: as you noted elsewhere, the execution context in which the player dies keeps on chugging. 

The band-aid for this is to replace the line that threw the error with something like

```
G.player.health.heal_or_harm(-1)
if G.just_died:
    return
for character in self.room.npcs:
    ...
```

But I'd prefer we figure out how to fix the problems with resetting more generally. The best I can come up with is exceptions. Anyone got a better idea? @xmunoz @abjose

**Comment by abjose** [comment]
Some random ideas:
- delay the restart in some way (like by appending it to the event queue? I haven't looked into how the event system works)
- do the G.just_died thing (wouldn't this be a similar amount of work to the exception version? though currently the just_died flag is being reset by the restart... and I guess without an exception we would have to be sure to add `return`s all the way back up the stack)
- have "run-specific globals"? Like each run / life has, in addition to the "persistent" globals, a set of global variables that are unique to that run and won't be modified by future or past runs (this is really more of a fix for the G.just_died flag being cleared by the future run)
- a simpler version of the above would be to assign an ID to each run and track the current run in a global variable
- MAKE EVERY RUN RECURSIVE (this would be good thematically and we could add fun flavor text for when the player exceeds the max recursion depth)

But I don't think any of these are better or easier.

**Comment by flosincapite** [comment]
> Some random ideas:
> 
>     * delay the restart in some way (like by appending it to the event queue? I haven't looked into how the event system works)
> 
>     * do the G.just_died thing (wouldn't this be a similar amount of work to the exception version? though currently the just_died flag is being reset by the restart... and I guess without an exception we would have to be sure to add `return`s all the way back up the stack)
> 
>     * have "run-specific globals"? Like each run / life has, in addition to the "persistent" globals, a set of global variables that are unique to that run and won't be modified by future or past runs (this is really more of a fix for the G.just_died flag being cleared by the future run)
> 
>     * a simpler version of the above would be to assign an ID to each run and track the current run in a global variable
> 
>     * MAKE EVERY RUN RECURSIVE (this would be good thematically and we could add fun flavor text for when the player exceeds the max recursion depth)

This is really cute! I think this will actually work, too.

> 
> 
> But I don't think any of these are better or easier.

**Comment by xmunoz** [comment]
> * MAKE EVERY RUN RECURSIVE 

I hate this idea, considering we don't currently have any tests. This could easily turn into a recipe for disaster.

**Comment by flosincapite** [comment]
> > * MAKE EVERY RUN RECURSIVE
> 
> I hate this idea, considering we don't currently have any tests. This could easily turn into a recipe for disaster.

Haha, fair point. 

So I just tested this out. The recursion solution does actually work.

@xmunoz, to your point, there is actually something virtuous about this solution from a testing perspective. It ensures that, each time the game restarts, all global state is completely cleaned up AND the call stack from the previous run will not pollute the current one. Using exceptions would accomplish the same thing, of course. Do you feel that one of these solutions is easier to test or control?

**Comment by abjose** [comment]
The recursion one was just supposed to be a joke! Wouldn't it make for giant stack traces?

**Comment by flosincapite** [comment]
Only a little giant :-p

**Comment by flosincapite** [comment]
So I like some of these solutions to address various other issues. Unfortunately, none of them will solve the problem of functions continuing to execute after a reset--at least, not without some refactoring. In the long term, it would be useful for this game to have a proper ECS, which would (among other things) help separate state updates from input handling. I don't see us doing that anytime soon. For now, using exceptions is easy to implement and relatively safe.

**Comment by abjose** [comment]
Sounds good to me!

**Comment by abjose** [comment]
Still hitting this issue (in the same situation: dying from being pelted with debris) after rebasing on the Exception PR! Will look into it more soon.

**Comment by flosincapite** [comment]
Hmm. I just tested at `main` (and died from debris). I didn't get this exception, even when the player was definitely not the last character to get pelted. How are you running the game?

**Comment by abjose** [comment]
Sorry, I was wrong - the smoker actually dies!

Abbreviated Logs:

```
random command: attack smoker

You hit smoker for 1 damage!

Smoker makes no hostile motion.

[...]

random command: attack smoker

You hit smoker for 8 damage!

Smoker makes no hostile motion.

[...]

random command: look

An idol with ruby eyes[...]

Smoker puffs mellowly on a cigarette, looking extremely fly.

The smoke whirls for a moment on the altar as though set there for an offering. The idol's ruby eyes seem to blaze[...]

All beings present are pelted with debris.

Health damaged by 1.

The smoker glances placidly around the environs. "Until the next time around, I guess." With a final nod, he breathes a perfect wreath of smoke, which dissipates solemnly.
```
then
```python
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/BOSDYN/ajose/Downloads/cronenbroguelike/cronenbroguelike/__main__.py", line 24, in <module>
    game.main()
  File "/home/BOSDYN/ajose/Downloads/cronenbroguelike/cronenbroguelike/game.py", line 113, in main
    adventurelib._handle_command("random")
  File "/home/BOSDYN/ajose/.local/lib/python3.6/site-packages/adventurelib.py", line 600, in _handle_command
    func(**args)
  File "<string>", line 4, in wrapped
  File "/home/BOSDYN/ajose/Downloads/cronenbroguelike/cronenbroguelike/commands.py", line 92, in random_action
    adventurelib._handle_command(final_command)
  File "/home/BOSDYN/ajose/.local/lib/python3.6/site-packages/adventurelib.py", line 600, in _handle_command
    func(**args)
  File "<string>", line 4, in wrapped
  File "/usr/lib/python3.6/contextlib.py", line 88, in __exit__
    next(self.gen)
  File "/home/BOSDYN/ajose/Downloads/cronenbroguelike/engine/globals.py", line 91, in poll_events
    event.execute()
  File "/home/BOSDYN/ajose/Downloads/cronenbroguelike/cronenbroguelike/rooms.py", line 152, in execute
    for character in self.room.characters:
RuntimeError: Set changed size during iteration
```

so I think in this case just iterating backwards through the list of characters might do it?

**Comment by xmunoz** [comment]
Was there a PR or commit that fixed this issue?

**Comment by flosincapite** [comment]
Sorry, @abjose, I had not noticed your comment from last week. Yeah, this is a real bug and I think you opened a PR at some point addressing it. I will try to recreate that PR. Sorry for being a blithe and curmudgeonly reviewer!

**Comment by flosincapite** [comment]
Wait, I am a fool, I did merge your PR.

**Comment by flosincapite** [comment]
Fixed by #78

**Comment by abjose** [comment]
Sorry @xmunoz, I should have linked the PRs - this was fixed by a combo of https://github.com/Cronenbrogues/cronenbroguelike/pull/78 which Cory linked and https://github.com/Cronenbrogues/cronenbroguelike/pull/80

@flosincapite you are neither blithe nor curmudgeonly, but lithe and a smidgen twee (oh god)
